Sources/Datadog/DatadogCore/Context/AppStatePublisher.swift

```diff
@@ -0,0 +1,145 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-2020 Datadog, Inc.
+ */
+
+import Foundation
+
+internal final class AppStatePublisher: ContextValuePublisher {
+    typealias Snapshot = AppStateHistory.Snapshot
+
+    /// The default publisher queue.
+    private static let defaultQueue = DispatchQueue(
+        label: "com.datadoghq.app-state-publisher",
+        target: .global(qos: .utility)
+    )
+
+    /// The initial history value.
+    var initialValue: AppStateHistory
+
+    /// The notification center where this publisher observes following `UIApplication` notifications:
+    /// - `.didBecomeActiveNotification`
+    /// - `.willResignActiveNotification`
+    /// - `.didEnterBackgroundNotification`
+    /// - `.willEnterForegroundNotification`
+    private let notificationCenter: NotificationCenter
+
+    /// The date provider for the Application state snapshot timestamp.
+    private let dateProvider: DateProvider
+
+    /// The queue used to serialise access to the `history` and
+    /// to publish the new history.
+    private let queue: DispatchQueue
+
+    /// The current application state history.
+    ///
+    /// To mutate in the `queue` only.
+    private var history: AppStateHistory
+
+    /// The receiver for publishing the state history.
+    ///
+    /// To mutate in the `queue` only.
+    private var receiver: ContextValueReceiver<AppStateHistory>?
+
+    /// Creates a Application state publisher for publishing application state
+    /// history.
+    ///
+    /// - Parameters:
+    ///   - initialState: The initial application state.
+    ///   - queue: The queue for publishing the history.
+    ///   - dateProvider: The date provider for the Application state snapshot timestamp.
+    ///   - notificationCenter: The notification center where this publisher observes `UIApplication` notifications.
+    init(
+        initialState: AppState,
+        queue: DispatchQueue = AppStatePublisher.defaultQueue,
+        dateProvider: DateProvider = SystemDateProvider(),
+        notificationCenter: NotificationCenter = .default
+    ) {
+        let snapshot = Snapshot(
+            state: initialState,
+            date: dateProvider.now
+        )
+
+        let initialValue = AppStateHistory(
+            initialSnapshot: snapshot,
+            recentDate: snapshot.date
+        )
+
+        self.initialValue = initialValue
+        self.history = initialValue
+        self.queue = queue
+        self.dateProvider = dateProvider
+        self.notificationCenter = notificationCenter
+    }
+
+    func publish(to receiver: @escaping ContextValueReceiver<AppStateHistory>) {
```

Comment:
nit/ The more often I see this in different publishers, the more I think it should be called `startPublishing(to:)` to better express what it should do 💭. Just a note, I'm totally fine with keeping existing name (after all, it is well explained in the protocol comment).

Reply:
That would work, current name also correctly reflect what it should do, at the protocol level, IMO. `publisher.publish(to receiver:)` Lets keep it as is for now 👍